src/defrag.c

```diff
@@ -844,76 +882,434 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-            listNode *head = listFirst(db->defrag_later);
-
-            /* Move on to next key */
-            if (defrag_later_current_key) {
-                serverAssert(defrag_later_current_key == head->value);
-                listDelNode(db->defrag_later, head);
-                defrag_later_cursor = 0;
-                defrag_later_current_key = NULL;
-            }
+    while (defrag_later && listLength(defrag_later) > 0) {
+        listNode *head = listFirst(defrag_later);
+        sds key = head->value;
+        dictEntry *de = kvstoreDictFind(ctx->kvstate.kvs, ctx->kvstate.slot, key);
 
-            /* stop if we reached the last one. */
-            head = listFirst(db->defrag_later);
-            if (!head) return 0;
+        long long key_defragged = server.stat_active_defrag_hits;
+        bool timeout = (defragLaterItem(de, &defrag_later_cursor, endtime, ctx->db->id) == 1);
+        if (key_defragged != server.stat_active_defrag_hits) {
+            server.stat_active_defrag_key_hits++;
+        } else {
+            server.stat_active_defrag_key_misses++;
+        }
 
-            /* start a new key */
-            defrag_later_current_key = head->value;
-            defrag_later_cursor = 0;
+        if (timeout) break;
+
+        if (defrag_later_cursor == 0) {
+            // the item is finished, move on
+            listDelNode(defrag_later, head);
         }
 
-        /* each time we enter this function we need to fetch the key from the dict again (if it still exists) */
-        dictEntry *de = kvstoreDictFind(db->keys, slot, defrag_later_current_key);
-        key_defragged = server.stat_active_defrag_hits;
+        if (++iterations > 16 || server.stat_active_defrag_hits - prev_defragged > 512 ||
+            server.stat_active_defrag_scanned - prev_scanned > 64) {
+            if (getMonotonicUs() > endtime) break;
+            iterations = 0;
+            prev_defragged = server.stat_active_defrag_hits;
+            prev_scanned = server.stat_active_defrag_scanned;
+        }
+    }
+
+    return (!defrag_later || listLength(defrag_later) == 0); // Done/True if nothing left
+}
+
+
+// This helper function handles most of the work for iterating over a kvstore.  'privdata', if
+// provided, MUST begin with 'kvstoreIterState' and this part is automatically updated by this
+// function during the iteration.
+static bool defragStageKvstoreHelper(monotime endtime,
+                                     kvstore *kvs,
+                                     dictScanFunction scan_fn,
+                                     kvstoreHelperPreContinueFn precontinue_fn,
+                                     const dictDefragFunctions *defragfns,
+                                     void *privdata) {
+    static kvstoreIterState state; // STATIC - this persists
+    if (endtime == 0) {
+        // Starting the stage, set up the state information for this stage
+        state.kvs = kvs;
+        state.slot = -2;
```

Comment:
```suggestion
        state.slot = KVSTORE_ITERATOR_SLOT_INIT;
```

Or something to make this clearer instead of a magic number.

Reply:
I confess, I didn't like the arbitrary number either.  But we were already using an arbitrary -1, so I figured what the heck.  I'll clean it up.